### ISSUE TITLE
New rule proposal: default-case-last

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12665]
**Please describe what the rule should do:**

Enforce `default` clauses in `switch` statements to be last.

**What category of rule is this? (place an "X" next to just one item)**

[X] Suggests an alternate way of doing something (suggestion)

**Provide 2-3 code examples that this rule will warn about:**

```js
switch (foo) {
  case 1:
    bar();
    break;
  default:
    baz();
    break;
  case 2:
    quux();
    break;    
}

switch (foo) {
  default:
    doSomethingForNonZero();
    // fallthrough
  case 0:
    doSomethingForAll();
    break;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule enforces a common convention. 

Even if a non-last `default` might make sense for a specific situation, it's better to refactor the code to a more understandable version.

A non-last `default` is, I believe, unexpected for readers. A reader might not know if it is a valid code at all, and what's the actual behavior (does the evaluation stop when it finds `default` in the middle, does it execute the `default` block and then continue searching, etc.).

This rule applies only to `switch` statements that already have `default` clauses. It doesn't enforce the existence of `default` clauses (there is the `default-case` rule for that purpose). 

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by platinumazure** [comment]
I didn't even know this was allowed in JS. Yes, we should definitely create a rule for enforcing this!

**Comment by mdjermanovic** [comment]
I'm working on this.

**Comment by shirohana** [comment]
Hi there,

I found that the rule `default-case-last` hasn't be linked correctly in https://eslint.org/docs/rules/default-case-last but shows in https://eslint.org/docs/rules/#best-practices, is it in a plan to be update or just a bug (´・ω・｀)?

**Comment by kaicataldo** [comment]
This is a bug in our website generator for prerelease versions. Sorry about that!
